### ISSUE TITLE
[Feature RFC]: cut sstables on a row boundaty

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#13091]
*Installation details*
HEAD: b6e427551193fd9a06d12e2c1de832e51a26ac00

*Problem*
Today scylla doesn't break a partition when it writes a new sstable even if this violates compaction strategy sstable size limitations.

As a result a huge partition can result in a huge sstable which will have a huge negative effect across the board: starting with a disk space amplification during the compaction and ending in huge RAM requirements due to pumped up promoted index.

*Solution*
Split sstables on a row boundary and respect compaction strategy sstable size configuration.

*Why it will work?*
Even today the same partition can be in many different sstables and `scylla` has all the machinery to read its data.

*Challenges*
There is a hard coded assumption that sstables in the same run are disjoint in compaction strategies today. We would need to allow them to share a single token (if sstable is cut on a row boundary) in order to implement this RFC.

**Comment by vladzcloudius** [comment]
@raphaelsc FYI

**Comment by raphaelsc** [comment]
we already support splitting on row boundary, but it will cause memory pressure today. note it doesn't cause memory with splitting on partition boundary because we have a read mechanism which is token aware, but not position in partition aware.

**Comment by raphaelsc** [comment]
each reader uses a buffer size of 128k. if I split large partition of 1TB, now I have 1k SSTables spanning a token T. If I read token T, I have to read from 1k SSTables in parallel. That's 128M per each read. Is this acceptable?

**Comment by raphaelsc** [comment]
Commit that added this functionality (not wired anywhere yet) https://github.com/scylladb/scylladb/commit/e2ccafbe38be8515523644a43534f596ee1db66e

**Comment by fee-mendes** [comment]
I think it all makes sense, and https://github.com/scylladb/scylladb/issues/12995 should be definitely implemented to avoid a huge partition altogether. Then, this and https://github.com/scylladb/scylladb/issues/13076 all become "nice to haves", just in case.

**Comment by vladzcloudius** [comment]
> each reader uses a buffer size of 128k. if I split large partition of 1TB, now I have 1k SSTables spanning a token T. If I read token T, I have to read from 1k SSTables in parallel. That's 128M per each read. Is this acceptable?

Is loading 8GB of compression info into the RAM for a huge sstable acceptable?

I think not and not is a requirement of allocation 128MB or RAM for a single read.
Did you consider *not* to read them all in such a huge concurrency and instead have an unlimited read concurrency? 
@raphaelsc solution for the problem you have described is quite trivial.

**Comment by raphaelsc** [comment]
> > we already support splitting on row boundary,
> 
> Really? I didn't know what. Since when?
> 

Please let's calm down. This won't get us anywhere. It's clear to all of us that large partitions must be split. And we'll get there. Like I said above, the commit with machinery was added, but not wired yet.

**Comment by raphaelsc** [comment]
> > each reader uses a buffer size of 128k. if I split large partition of 1TB, now I have 1k SSTables spanning a token T. If I read token T, I have to read from 1k SSTables in parallel. That's 128M per each read. Is this acceptable?
> 
> Is loading 8GB of compression info into the RAM for a huge sstable acceptable?

IIRC, compression metadata is proportional to the amount of data, therefore it won't get smaller after splitting.

> 
> I think not and not is a requirement of allocation 128MB or RAM for a single read. Did you consider _not_ to read them all in such a huge concurrency and instead have a limited read concurrency? @raphaelsc solution for the problem you have described is quite trivial.

@denesb please clarify further. But given the status quo, our reader machinery currently must open all SSTables that span a token T for it to produce the compacted rows in the correct order. I don't think limiting concurrency for a combined reader will solve this problem. It's k-way merge. Each SSTable being read from will keep a buffer, limited in size, with fragments spanning current token being read, and then we'll combine the results to produce the output fragments.

**Comment by raphaelsc** [comment]
what will effectively limit read concurrency is what I proposed in the doc: https://docs.google.com/document/d/1Gb2o_8VeDbLXX-p7Qb7bql3_yP2qMIpxqReL6H0bbhI/edit#heading=h.n8t0kiwn1xo6

Given that large partition split won't overlap in row range, we can read the fragments incrementally. That's basically what we intend and agreed to do.

**Comment by denesb** [comment]
> @denesb please clarify further. But given the status quo, our reader machinery currently must open all SSTables that span a token T for it to produce the compacted rows in the correct order. I don't think limiting concurrency for a combined reader will solve this problem. It's k-way merge. Each SSTable being read from will keep a buffer, limited in size, with fragments spanning current token being read, and then we'll combine the results to produce the output fragments.

The combined reader has to know the current position of each participating read. For those that are on the same position, the combined reader has to see the next fragment from all of them. This can be done sequentially although for this to work, we'd need quite intrusive changes in our reader pipeline. We have buffering on every level, so currently it doesn't matter if you read parallelly or sequentially, you'd end up with the same amount of memory in the end (although it does help with transient memory usage like index reads). Maybe all we need is a `set_buffer_size_recursive()`  method on the readers, setting buffer size to 0 on the entire stack and adjusting the readers to read exactly one fragment when this is so.

**Comment by denesb** [comment]
Note though, that since aab5954cf, user reads reading from 1K sstables, would likely get killed, only compaction could OOM the node.

**Comment by vladzcloudius** [comment]
> IIRC, compression metadata is proportional to the amount of data, therefore it won't get smaller after splitting.

Do we always keep a compression metadata of all sstables in the memory? Why?

**Comment by denesb** [comment]
> > IIRC, compression metadata is proportional to the amount of data, therefore it won't get smaller after splitting.
> 
> Do we always keep a compression metadata of all sstables in the memory? Why?

Yes we do. If we didn't, we would have to do exactly 2X as many I/O to read from sstables: 1 read to read the relevant compression info entry, and another to read the data file.

**Comment by vladzcloudius** [comment]
> > > IIRC, compression metadata is proportional to the amount of data, therefore it won't get smaller after splitting.
> > 
> > 
> > Do we always keep a compression metadata of all sstables in the memory? Why?
> 
> Yes we do. If we didn't, we would have to do exactly 2X as many I/O to read from sstables: 1 read to read the relevant compression info entry, and another to read the data file.

Why don't we read compression info in a lazy way, @denesb? Why do you need to keep a compression info of the sstable that you are not planning to read from?

**Comment by denesb** [comment]
> Why don't we read compression info in a lazy way, @denesb? Why do you need to keep a compression info of the sstable that you are not planning to read from?

Do we have sstables we don't plan to read from? What we read is determined by the user, we have no idea what the user will want to read. We plan to read from all sstables we have.

We could create a compression info cache where compression infos are loaded on first use. If we have enough memory we keep all in memory, if not we evict the least recently used to make room for a new one. But this would add a lot of additional complexity and a open the door for a whole new class of bugs. It would also add an unpredictable latency element, possibly destroying our 99th percentile. A read which has to load the compression-info element even for a single sstable would get very bad latency.

I think this is just not worth it, we choose an entirely different direction to support larger memory:disk ratios: moving data to S3. So I don't think it is worth to try hard improving a node's local memory:disk ratio, the node local disk is soon going to be just a cache itself anyway.

**Comment by vladzcloudius** [comment]
> Do we have sstables we don't plan to read from? What we read is determined by the user, we have no idea what the user will want to read. We plan to read from all sstables we have.

We can (and I suggested we should) plan *not* to read from any sstable until we are asked to. I assumed you were familiar with the "lazy way" term. ;)

> We could create a compression info cache where compression infos are loaded on first use. If we have enough memory we keep all in memory, if not we evict the least recently used to make room for a new one. But this would add a lot of additional complexity and a open the door for a whole new class of bugs. It would also add an unpredictable latency element, possibly destroying our 99th percentile. A read which has to load the compression-info element even for a single sstable would get very bad latency.

I think you are exaggerating to say the least - we have a much heavier component we read on-demand today - the index, and we are doing just fine.
And not caching it doesn't "destroy P99" in a general case.
So, reading the compression info component which is an order of magnitude smaller wouldn't destroy anything either:

```
$ ll ~/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/*-CompressionInfo.db                                                                                 
-rw-rw-r-- 1 vladz vladz 252K Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-10-big-CompressionInfo.db
-rw-rw-r-- 1 vladz vladz  57K Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-8-big-CompressionInfo.db
-rw-rw-r-- 1 vladz vladz 127K Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-9-big-CompressionInfo.db
$ ll ~/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/*-Index.db                                                                                           
-rw-rw-r-- 1 vladz vladz 9.4M Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-10-big-Index.db
-rw-rw-r-- 1 vladz vladz 2.2M Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-8-big-Index.db
-rw-rw-r-- 1 vladz vladz 4.8M Mar 16 10:25 /home/vladz/.ccm/scylla-1/node1/data/keyspace1/standard1-64b7e4b0be1411ed92b9a2cff83cc434/me-9-big-Index.db
```

> I think this is just not worth it, we choose an entirely different direction to support larger memory:disk ratios: moving data to S3. So I don't think it is worth to try hard improving a node's local memory:disk ratio, the node local disk is soon going to be just a cache itself anyway.

How is S3 support is even relevant to this to begin with?
And, second, what you write makes little sense in general.
It will always make a lot of sense to improve our RAM (!!) usage regardless of how we use a local disk.

To remind you the context of my original question: the fact that we needlessly load a compression info of all sstables in advance is going to cause Scylla to OOM in case of a very large partition: https://github.com/scylladb/scylladb/issues/13091#issuecomment-1457216168

So, unless we resolve this no "S3" is going to help us.

And I also think you are a little confused about what's planned to be implemented in the context of "S3" support. What you are describing is usually called "tiered storage" and AFAIK no plans to implement a tiered storage exist yet.

**Comment by mykaul** [comment]
> And I also think you are a little confused about what's planned to be implemented in the context of "S3 support". Or maybe I don't know something you do. What you are describing is usually called "tiered storage" and AFAIK no plans to implement a tiered storage exist yet.

Inaccurate - we have plans to implement both caching and tiering.

**Comment by vladzcloudius** [comment]
> > And I also think you are a little confused about what's planned to be implemented in the context of "S3 support". Or maybe I don't know something you do. What you are describing is usually called "tiered storage" and AFAIK no plans to implement a tiered storage exist yet.
> 
> Inaccurate - we have plans to implement both caching and tiering.

I'd love to hear more about it then, @mykaul.
Because every time I was asking about it I was getting a "definitely no" answer.

However, as I mentioned, this is completely irrelevant to the subject of this RFC.

**Comment by denesb** [comment]
> We can (and I suggested we should) plan not to read from any sstable until we are asked to. That's what a "lazy way" stands for ;)

I know what lazy means. But with sstables we are quite eager: we pre-read all but the largest components into memory. We only go to to disk for the index and data components.

> I think you are exaggerating to say the least - we have a much heavier component we read on-demand today - the index, and we are doing just fine.

We don't eve read the entire index. We usually read a single page, as determined by the summary.

> And not caching it doesn't "destroy P99" in a general case.
> So, reading the compression info component which is an order of magnitude smaller wouldn't destroy anything either:

I meant reading the entire compression-info into memory on first use. Just reading the required entry of course wouldn't destroy the 99'th, however it would raise the now average 2 IO/read -> 3 IO/read, a 50% increase. This probably wouldn't translate to a 50% latency increase but would be noticeable to be sure.

**Comment by denesb** [comment]
> How is S3 support is even relevant to this to begin with?

Actually, s3 support has more to do with this than you might think. Currently S3 (as presented in https://github.com/scylladb/scylladb/pull/12523) is not helping our memory:disk ration at all. It just extends our disk storage but does nothing to help us improve memory:disk, because even for sstables stored in S3 the usual in-memory components are still kept in memory. So they don't take space on the local disk, but they do take space in memory. So we will need a solution to reduce the in-memory footprint of sstables, if we want to really benefit from offloaded sstables. And whatever we come up with, we can apply to local sstables too (if we decide we want to).

**Comment by denesb** [comment]
> Clearly this approach backfires - hence this whole discussion.

I wouldn't say backfires. Today we aim for minimum latency at the cost of memory consumption. We can reduce the memory consumption by sacrificing latency. No matter how sophisticated of a caching system we come up with, some reads are going to miss and those are going to suffer. How much? We need benchmarks to tell that.

But like I said in my other comment, we are working towards tiered storage and I think it makes sense to allow adding local sstables to the cold tier too. In the cold tier, we could use caching or not even that -- go all in in optimizing for minimum memory consumption instead of latency.

**Comment by vladzcloudius** [comment]
> > Clearly this approach backfires - hence this whole discussion.
> 
> I wouldn't say backfires. 

Please, see the referenced post to see what I refer by "backfires" and it clearly backfires.

> Today we aim for minimum latency at the cost of memory consumption. We can reduce the memory consumption by sacrificing latency. No matter how sophisticated of a caching system we come up with, some reads are going to miss and those are going to suffer. How much? We need benchmarks to tell that.

This RFC is not about performance - it's about stability. 
It's not about optimizing a memory consumption - it's about avoiding OOM.

> 
> But like I said in my other comment, we are working towards tiered storage and I think it makes sense to allow adding local sstables to the cold tier too. In the cold tier, we could use caching or not even that -- go all in in optimizing for minimum memory consumption instead of latency.

And like I also said - this is completely irrelevant. Please, let's concentrate on solving the problem in question. It's generic and no tired or non-tired storage is going to solve it.

**Comment by vladzcloudius** [comment]
@raphaelsc @denesb I've created a separate issue for sstables' parts caching: https://github.com/scylladb/scylladb/issues/13219

**Comment by denesb** [comment]
Indeed the s3 effort is only going to help here if it comes with tiered storage implementation, and therefore it solves the problem of detaching disk from memory (the memory:disk ratio).

Considering that according to @mykaul we are planning for tackling tiered storage, I think it is worth prioritizing this. Normally we are quite reluctant to expend a lot of effort to make ScyllaDB well behaved on anti-pattern workloads, of which huge partitions is definitely one of. Yet we did expend a lot of effort in the past years to support large partitions, because as we discovered, users "love" them, they are a fact of life.
So I think it is worth to look into supporting using compression info and possibly other currently in-memory components from the disk, as it kills two birds with one stone:
* it allows for ScyllaDB to continue to function in the face of truly enormous partitions
* it is a requirement for tiered storage and S3 anyway

Let's move the discussion on how to do it to the new issue @vladzcloudius opened: #13219.

**Comment by vladzcloudius** [comment]
Right. A few clarifications however (see inline):

> Indeed the s3 effort is only going to help here if it comes with tiered storage implementation, and therefore it solves the problem of detaching disk from memory (the memory:disk ratio).
> 
> Considering that according to @mykaul we are planning for tackling tiered storage, I think it is worth prioritizing this. 

"This" stands for this GH issue proposal I guess, right?

> Normally we are quite reluctant to expend a lot of effort to make ScyllaDB well behaved on anti-pattern workloads, of which huge partitions is definitely one of. 

Here by "well-behaved" we (I assume) refer "not crash". Good performance is such a situation is not expected at all.


> Yet we did expend a lot of effort in the past years to support large partitions, because as we discovered, users "love" them, they are a fact of life. 

Sometimes it's not about "love" - it happens due to bugs in the application or some ill-behaved user who's out of control of the Operator.

> So I think it is worth to look into supporting using compression info and possibly other currently in-memory components from the disk, as it kills two birds with one stone:
> 
>     * it allows for ScyllaDB to continue to function in the face of truly enormous partitions
> 
>     * it is a requirement for tiered storage and S3 anyway
> 
> 
> Let's move the discussion on how to do it to the new issue @vladzcloudius opened: #13219.

**Comment by denesb** [comment]
> > Indeed the s3 effort is only going to help here if it comes with tiered storage implementation, and therefore it solves the problem of detaching disk from memory (the memory:disk ratio).
> > Considering that according to @mykaul we are planning for tackling tiered storage, I think it is worth prioritizing this.
> 
> "This" stands for this GH issue proposal I guess, right?

Here I meant improving the memory:disk ratio, or https://github.com/scylladb/scylladb/issues/13219.
We already decided this present issue is worth doing and @raphaelsc started working on it, it just god preemted with something else.

> > Normally we are quite reluctant to expend a lot of effort to make ScyllaDB well behaved on anti-pattern workloads, of which huge partitions is definitely one of.
> 
> Here by "well-behaved" we (I assume) refer "not crash". Good performance in such a situation is not expected at all.

Yes, we should not be reluctant at all to achieve the (quite low) bar of not crashing.
